addons/hr_attendance/models/hr_employee.py

```diff
@@ -187,8 +187,8 @@ def _attendance_action_change(self):
                 'employee_id': self.id,
                 'check_in': action_date,
             }
-            return self.env['hr.attendance'].create(vals)
-        attendance = self.env['hr.attendance'].search([('employee_id', '=', self.id), ('check_out', '=', False)], limit=1)
+            return self.env['hr.attendance'].sudo().create(vals)
+        attendance = self.env['hr.attendance'].sudo().search([('employee_id', '=', self.id), ('check_out', '=', False)], limit=1)
```

Comment:
do you know why the extra `sudo` are needed here, considering that the only calls are already on a sudo'ed employee env?
https://github.com/odoo/odoo/blob/master/addons/hr_attendance/models/hr_employee.py#L158-L169

Perhaps the tests could be fixed instead to call `_attendance_action()`?

Reply:
> do you know why the extra `sudo` are needed here, considering that the only calls are already on a sudo'ed employee env? https://github.com/odoo/odoo/blob/master/addons/hr_attendance/models/hr_employee.py#L158-L169
> 
> Perhaps the tests could be fixed instead to call `_attendance_action()`?

It looks like the problem doesn't come from the test using `_attendance_action_change` but from the `with_user` within the method `_attendance_action`, which looses the sudo:

```
Traceback (most recent call last):
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/addons/hr_attendance/tests/test_hr_attendance_process.py", line 47, in test_checkin_self_without_pin
    employee.with_user(self.user).attendance_manual({}, entered_pin=None)
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/addons/hr_attendance/models/hr_employee.py", line 147, in attendance_manual
    return self._attendance_action(next_action)
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/addons/hr_attendance/models/hr_employee.py", line 167, in _attendance_action
    modified_attendance = employee.with_user(employee.user_id)._attendance_action_change()
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/addons/hr_attendance/models/hr_employee.py", line 190, in _attendance_action_change
    return self.env['hr.attendance'].create(vals)
  File "<decorator-gen-278>", line 2, in create
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/api.py", line 409, in _model_create_multi
    return create(self, [arg])
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/addons/hr_attendance/models/hr_attendance.py", line 275, in create
    res = super().create(vals_list)
  File "<decorator-gen-67>", line 2, in create
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/api.py", line 410, in _model_create_multi
    return create(self, arg)
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/addons/base/models/ir_fields.py", line 613, in create
    recs = super().create(vals_list)
  File "<decorator-gen-16>", line 2, in create
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/api.py", line 410, in _model_create_multi
    return create(self, arg)
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/models.py", line 4120, in create
    self.check_access_rights('create')
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/models.py", line 3683, in check_access_rights
    return self.env['ir.model.access'].check(self._name, operation, raise_exception)
  File "<decorator-gen-35>", line 2, in check
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/home/odoo/src/odoo/master-ir-ui-view-remove-groups-id-for-backend-views-dle/odoo/addons/base/models/ir_model.py", line 1821, in check
    raise AccessError(msg)
odoo.exceptions.AccessError: You are not allowed to create 'Attendance' (hr.attendance) records.

This operation is allowed for the following groups:
        - Attendances/Officer : Manage all attendances
        - Attendances/User

Contact your administrator to request access if necessary.
```

I therefore moved the sudo that I put to this `with_user`, so the sudo are all kept within the method `_attendance_action`